### PR TITLE
Engine implement Configuration module

### PR DESCRIPTION
**Pull request by JavierBejMen**
|Related issue|
|---|
|#15939|

Adds the configuration module to the Engine.

The Engine is able to run without a configuration file, with default values hard coded. 

The configuration on startup can be modified through console parameters, a configuration file, and environment variables, described in order from most to least prevalent.

In runtime configuration can be queried and modified through the configuration API, though changes won’t take effect until restart as of now, aside from saving the new configuration.

## Changes
- Refactored CMDS for clarity and modularity
- Added conf module, with IConf interface and CliConf configuration driver using Cli11 library.

### CODE REVIEW

**Review by juliancnn — CHANGES_REQUESTED** [review]

File: src/engine/source/main.cpp [hunk 1]
```diff
@@ -1,747 +1,60 @@
...
-std::shared_ptr<CLI::App> configureCliApp()
+int main(int argc, char* argv[])
 {
-    auto app = std::make_shared<CLI::App>(
+    CLI::App_p app = std::make_shared<CLI::App>(
         "The Wazuh engine analyzes all the events received from agents, remote devices "
         "and Wazuh integrations. This integrated console application allows to manage "
-        "all the engine components.\n");
+        "all the engine components.");
+    // app->require_subcommand(1);
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/main.cpp:27]: Uncommented!

File: src/engine/source/cmds/src/start.cpp [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    std::string apiEndpoint;
+    int queueSize;
+    int threads;
+    std::string fileStorage;
+    int logLevel;
+    std::string logOutput;
+    std::string environment;
+};
+
+constexpr auto START_ENVIRONMENT = "environment/wazuh/0";
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/cmds/src/start.cpp:57]: deleted

File: src/engine/source/cmds/src/start.cpp [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        ->envname(ENGINE_ENVIRONMENT_ENV);
+
+    // Register callback
+    startApp->callback(
+        [app, options]()
+        {
+            auto confManager =
+                std::make_shared<conf::IConf<conf::CliConf>>(conf::CliConf(app));
+            runStart(confManager);
+        });
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/cmds/src/start.cpp:316]: Yes, to hold the variables binded to the Cli11 options

File: src/engine/source/conf/src/cliconf.cpp [hunk 4]
```diff
@@ -0,0 +1 @@
+#include <conf/cliconf.hpp>
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/conf/src/cliconf.cpp:1]: Moved implementation to cpp

File: src/engine/source/cmds/src/defaultSettings.hpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+constexpr auto ENGINE_LOG_OUTPUT_ENV = "WZE_LOG_OUTPUT";
+
+constexpr auto ENGINE_ENVIRONMENT = "environment/wazuh/0";
+constexpr auto ENGINE_ENVIRONMENT_ENV = "WZE_ENVIRONMENT";
+
+constexpr auto ENGINE_QUEUE_SIZE = 1000000;
+constexpr auto ENGINE_QUEUE_SIZE_ENV = "WZE_QUEUE_SIZE";
+
+constexpr auto ENGINE_LOG_LEVEL = 3;
+constexpr auto ENGINE_LOG_LEVEL_ENV = "WZE_LOG_LEVEL";
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/cmds/src/defaultSettings.hpp:29]: Setted as 2

File: src/engine/source/cmds/src/start.cpp [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        ->default_val(ENGINE_LOG_OUTPUT)
+        ->check(CLI::ExistingFile)
+        ->envname(ENGINE_LOG_OUTPUT_ENV);
+
+    // Server
+    // Endpoints
+    serverApp->add_option("--event_endpoint", options->eventEndpoint, "Sets the events server socket address.")
+        ->default_val(ENGINE_EVENT_SOCK)
+        ->envname(ENGINE_EVENT_SOCK_ENV);
+    serverApp->add_option("--api_endpoint", options->apiEndpoint, "Sets the API server socket address.")
```
Comment (LucioDonda) [inline src/engine/source/cmds/src/start.cpp:249]: we should unify all options, to `api_socket` or `api_endpoint` including help messages

**Review by LucioDonda — CHANGES_REQUESTED** [review]

File: src/engine/source/api/src/registry.cpp [hunk 7]
```diff
@@ -31,7 +31,7 @@ CommandFn Registry::getCallback(const std::string& command)
     return [command](const json::Json&)
     {
         return WazuhResponse {
-            json::Json {}, -1, fmt::format("Command \"{}\" not found", command)};
+            json::Json {"{}"}, -1, fmt::format("Command \"{}\" not found", command)};
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/api/src/registry.cpp:34]: done

File: src/engine/source/cmds/include/cmds/details/kbhit.hpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <sstream>
+#include <string>
+#include <unistd.h>
+#include <vector>
+
+namespace cmd::details
+{
+
+inline int kbhit()
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/cmds/include/cmds/details/kbhit.hpp:12]: done

File: src/engine/source/cmds/src/test.cpp [hunk 9]
```diff
@@ -1,10 +1,14 @@
...
+#include "cmds/test.hpp"
 
 #include <atomic>
+#include <chrono>
 #include <memory>
+#include <thread>
 
 #include <re2/re2.h>
 
+#include <cmds/details/kbhit.hpp>
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/cmds/src/test.cpp:10]: deleted

File: src/engine/source/cmds/src/start.cpp (hunk 6 above)
Comment (JavierBejMen) [reply src/engine/source/cmds/src/start.cpp:249]: changed to socket

File: src/engine/source/cmds/src/config.cpp [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        ->default_val("");
+    get_subcommand->callback([options]() { runGet(options->socketPath, options->name); });
+
+    auto save_subcommand = configApp->add_subcommand("save", "Persist the current configuration");
+    save_subcommand
+        ->add_option("path", options->path, "Path to save the configuration, empty to save in the default path")
+        ->default_val("");
+    save_subcommand->callback([options]() { runSave(options->socketPath, options->path); });
+
+    auto put_subcommand = configApp->add_subcommand("put", "Update a configuration option");
```
Comment (LucioDonda) [inline src/engine/source/cmds/src/config.cpp:110]: should we state somewhere that the server should be restarted for the actual change to take place?

**Review by LucioDonda — APPROVED** [review]
LGTM

**Review by JcabreraC — CHANGES_REQUESTED** [review]

File: src/engine/source/cmds/src/config.cpp (hunk 10 above)
Comment (JavierBejMen) [reply src/engine/source/cmds/src/config.cpp:110]: stated

**Review by JcabreraC — APPROVED** [review]
LGTM !

**Review by nbertoldo — APPROVED** [review]
LGTM

**Review by nmkoremblum — APPROVED** [review]
LGTM

**Pull request merged**